pkg/client/client.go

```diff
@@ -90,71 +101,95 @@ func (c *tlsBootstrapClientImpl) GetBootstrapToken(ctx context.Context) (string,
 	if err != nil {
 		return "", fmt.Errorf("failed to retrieve attested data from IMDS: %w", err)
 	}
+	c.logger.Info("retrieved IMDS attested data")
 
-	c.logger.Info("retrieving bootstrap token from TLS bootstrap token server")
-	tokenRequest := pb.TokenRequest{
+	c.logger.Debug("retrieving bootstrap token from TLS bootstrap server...")
+	tokenResponse, err := pbClient.GetToken(ctx, &pb.TokenRequest{
 		ResourceId:   instanceData.Compute.ResourceID,
-		Nonce:        nonce.Nonce,
+		Nonce:        nonce,
 		AttestedData: attestedData.Signature,
-	}
-	tokenReply, err := pbClient.GetToken(ctx, &tokenRequest)
+	})
 	if err != nil {
 		return "", fmt.Errorf("failed to retrieve a token: %w", err)
 	}
-	c.logger.Info("received token reply")
+	c.logger.Info("received new bootstrap token from TLS bootstrap server")
 
-	execCredential.APIVersion = "client.authentication.k8s.io/v1"
-	execCredential.Kind = "ExecCredential"
-	execCredential.Status.Token = tokenReply.Token
-	execCredential.Status.ExpirationTimestamp = tokenReply.Expiration
+	c.logger.Debug("generating new exec credential with bootstrap token...")
+	execCredentialWithToken, err := getExecCredentialWithToken(tokenResponse.GetToken(), tokenResponse.GetExpiration())
+	if err != nil {
+		return "", fmt.Errorf("unable to generate new exec credential with bootstrap token: %w", err)
+	}
+	c.logger.Info("generated new exec credential with bootstrap token")
 
-	execCredentialBytes, err := json.Marshal(execCredential)
+	execCredentialBytes, err := json.Marshal(execCredentialWithToken)
 	if err != nil {
 		return "", fmt.Errorf("failed to marshal execCredential")
 	}
+
 	return string(execCredentialBytes), nil
 }
 
-func loadExecCredential(logger *logrus.Logger) (*datamodel.ExecCredential, error) {
-	logger.WithField(kubernetesExecInfoVarName, os.Getenv(kubernetesExecInfoVarName)).Debugf("parsing %s variable", kubernetesExecInfoVarName)
-	kubernetesExecInfoVar := os.Getenv(kubernetesExecInfoVarName)
-	if kubernetesExecInfoVar == "" {
-		return nil, fmt.Errorf("%s variable not found", kubernetesExecInfoVarName)
-	}
+func getExecCredentialWithToken(token, expirationTimestamp string) (*datamodel.ExecCredential, error) {
+	if token == "" {
+		return nil, fmt.Errorf("token string is empty, cannot generate exec credential")
+	}
+	if expirationTimestamp == "" {
+		return nil, fmt.Errorf("token expiration timestamp is empty, cannot generate exec credential")
+	}
+	return &datamodel.ExecCredential{
+		APIVersion: "client.authentication.k8s.io/v1",
+		Kind:       "ExecCredential",
+		Status: datamodel.ExecCredentialStatus{
+			Token:               token,
+			ExpirationTimestamp: expirationTimestamp,
+		},
+	}, nil
+}
 
+func loadExecCredential() (*datamodel.ExecCredential, error) {
+	execInfo := os.Getenv(kubernetesExecInfoVarName)
+	if execInfo == "" {
+		return nil, fmt.Errorf("%s must be set to retrieve bootstrap token", kubernetesExecInfoVarName)
+	}
 	execCredential := &datamodel.ExecCredential{}
```

Comment:
prefer allocating on the stack vs. heap where possible. i.e., var `execCredential datamodel.ExecCredential`